rfcs/0045-zkapp-balance-data-in-archive.md

```diff
@@ -0,0 +1,292 @@
+  function `Archive_lib.Diff.Builder.breadcrumb_added`. The staged
+  ledger is contained in the breadcrumb argument, and the block
+  contains transactions, so the accounts affected by the block can be
+  queried from the staged ledger.
+
+- The same message can contain a list of records representing account
+  creation fees burned for the block, with fields for the public key
+  of the created account and the fee amount. That information can be
+  extracted from the scan state of the staged ledger in the breadcrumb.
+
+Types to modify:
+
+- The type `Mina_transaction_logic.Transaction_applied.Signed_command_applied`
+  contains the field `user_command` with a status, which will contain
+  the account creation fees.  Modify the nearby types
+  `Fee_transfer_applied` and `Coinbase_applied` to contain those
+  fees. We could use `With_status.t` for that purpose, or, since those
+  commands always succeed, add new fields to those types to hold the
+  account creation fee list.
+
+Removing the information from transaction statuses changes the
+structure of blocks, which use a versioned type. While the version
+number need not be changed, because the new type will be deployed at a
+hard fork, the change affects the serialization of blocks, so a new
+`Bin_prot.t` layout will be needed for `External_transition.t`.
+(The change in PR #10684, which removes the validation callback from
+blocks (serialized as the unit value), would also require a new
+layout, independently of the changes here.)
+
+## Changes to the archive processor
+
+The archive processor will need to be updated to add the account information that's
+changed for each block. There is no existing code to add entries to the
+`blocks_zkapps_commands` join table, it needs to be added.
+
+Because the `Breadcrumb_added` message will contain account creation fee
+information, the processor will no longer require the temporizing hack
+to calculate that information. The creation fee information will no
+longer be written to join tables, instead it will be written to the
+new `accounts_created` table.
+
+### Changes to archive blocks
+
+For extensional blocks (the type `Extensional.t`), a field `zkapps_cmds` needs to be added,
+and the function `add_from_extensional` needs to use that field to populate the tables
+`zkapp_commands` and `blocks_zkapp_commands`. Also, there needs to be a field `accounts`
+with the account information used to populate the new `accounts_accessed` table.
+
+Likewise, precomputed blocks (type `External_transition.Precomputed_block.t`) will need
+an `accounts` field with account information.
+
+These archive block types do not contain version information in their JSON serialization.
+Therefore, blocks exported with changed types will require code using the same types, and
+old exported blocks will not be importable by new code.
+
+## Changes to the replayer
+
+The replayer currently verifies balances, nonces, and account creation fees after
```

Comment:
Do we want to use the transaction logic on a 'db-backed ledger' implementation? We can also store merkle paths if we want to validate ledger hashes too (if that's worth the additional storage cost)?

Reply:
Sorry, not following what you're suggesting.

The current replayer validates ledger hashes per-block, using the hashes stored in the `blocks` table, against a ledger created with `Ledger.create`, which is db-backed.

Reply:
I was suggesting that we could use the database as a ledger (if we have the merkle paths), since it now has all the information that we need to apply a transaction, rather than constructing a ledger by replaying from the beginning of time.

Reply:
Yes, that's possible, but I think out of scope for this RFC, whose goal is to enable storing all the zkApp data in the archive db.